### PR TITLE
`resolveSymbol` for a reexport from a wrapped asset

### PR DESCRIPTION
**Pull request by mischnic**
# ↪️ Pull Request

1. Closes https://github.com/parcel-bundler/parcel/issues/4364

When resolve a symbol from a wrapped asset failed (the wrapping in hoist, not in concat), it didn't fallback to `exports.symbol` but threw a build error.

Now `symbols.clear()` (sets `asset.symbols = null`) signifies: "static analysis failed, bail out" and this way `BundleGraph#resolveSymbol` can use that information (previously using `meta.isCommonJS` was used in link.js in a similar situation). 

2. Fixes @wbinnssmith's icon problem (which popped up in the tests after fixing 1.)

`export {a as b} from "./foo"` resulted in resolveSymbol returning the `asset` "foo" but the `exportSymbol` "b" (and `getExportedSymbols()` also relied on this).
Now, `getExportedSymbols` returns `exportAs` as well.

### CODE REVIEW

**Comment by parcel-benchmark** [comment]
## Benchmark Results
<details><summary>Kitchen Sink ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 6.20s | -234.00ms |
| Cached | 2.27s | -20.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/parcel.d5807e82.webp | 102.94kb | +0.00b | 140.00ms | -145.00ms 🚀 |
| dist/modern/parcel.d5807e82.webp | 102.94kb | +0.00b | 140.00ms | -144.00ms 🚀 |
| dist/legacy/kitchen-sink.679578a5.js | 1.16kb | +0.00b | 1.37s | -171.00ms 🚀 |
| dist/modern/kitchen-sink.90de237b.js | 1.16kb | +0.00b | 554.00ms | -984.00ms 🚀 |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/parcel.d5807e82.webp | 102.94kb | +0.00b | 184.00ms | -39.00ms 🚀 |
| dist/modern/parcel.d5807e82.webp | 102.94kb | +0.00b | 183.00ms | -39.00ms 🚀 |
| dist/legacy/kitchen-sink.679578a5.js | 1.16kb | +0.00b | 379.00ms | -97.00ms 🚀 |
| dist/modern/kitchen-sink.90de237b.js | 1.16kb | +0.00b | 379.00ms | -96.00ms 🚀 |
| dist/legacy/index.html | 709.00b | +0.00b | 380.00ms | -96.00ms 🚀 |
| dist/modern/index.html | 709.00b | +0.00b | 379.00ms | -95.00ms 🚀 |
| dist/legacy/styles.afb8e31a.css | 78.00b | +0.00b | 380.00ms | -95.00ms 🚀 |
| dist/modern/styles.cda8c91b.css | 78.00b | +0.00b | 379.00ms | -96.00ms 🚀 |

</p></details>

<details><summary>React HackerNews ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 9.55s | +208.00ms |
| Cached | 2.29s | -4.00ms |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/main/index.js | 66.80kb | +0.00b | 193.00ms | -39.00ms 🚀 |
| dist/module/index.js | 37.47kb | +0.00b | 192.00ms | -39.00ms 🚀 |

</p></details>

<details><summary>AtlasKit Editor 🚨</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | **FAILED** | -0.00ms |
| Cached | **FAILED** | -0.00ms |

#### Cold Bundles

*No bundles found, this is probably a failed build...*
#### Cached Bundles

*No bundles found, this is probably a failed build...*
</p></details>

<details><summary>Three.js x4 🚨</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | **FAILED** | -0.00ms |
| Cached | **FAILED** | -0.00ms |

#### Cold Bundles

*No bundles found, this is probably a failed build...*
#### Cached Bundles

*No bundles found, this is probably a failed build...*
</p></details>

**[Click here to view a detailed benchmark overview.](https://benchmark.parceljs.org/benchmark/264791866674774529)**

File: packages/core/core/src/BundleGraph.js [hunk 1]
```diff
@@ -808,42 +802,52 @@ export default class BundleGraph {
...
+        }
       }
     }
 
     return {
       asset,
       exportSymbol: symbol,
-      symbol: identifier,
-      loc: asset.symbols.get(symbol)?.loc,
+      symbol: identifier ?? (maybeFoundInDependencies ? null : undefined),
```
Comment (wbinnssmith) [inline packages/core/core/src/BundleGraph.js:827]: Generally I find relying on having a semantic difference between `null` and `undefined` values makes things unclear. Could we use another special value instead? Or is this difference documented somewhere?

File: packages/core/core/src/public/Symbols.js [hunk 2]
```diff
@@ -5,95 +5,148 @@ import type {
...
     }
     return false;
   }
 
   // $FlowFixMe
   [Symbol.iterator]() {
-    return this.#symbols[Symbol.iterator]();
+    return this.#value.symbols
+      ? this.#value.symbols[Symbol.iterator]()
+      : {
```
Comment (wbinnssmith) [inline packages/core/core/src/public/Symbols.js:51]: Maybe turn this into a function called `createEmptyIterator` which returns this object. Would make this clearer, and could be used in the same way this is used below too.

File: packages/core/core/src/BundleGraph.js (hunk 1 above)
Comment (mischnic) [reply packages/core/core/src/BundleGraph.js:827]: I agree that this is not ideal, but I don't know of a better way.
I've documented it here
https://github.com/parcel-bundler/parcel/blob/f520e215405a62de59ec34a7d089618663b9cb7e/packages/core/core/src/BundleGraph.js#L737-L740
I'll move it to `@parcel/types`

Comment (wbinnssmith) [reply packages/core/core/src/BundleGraph.js:827]: Could we use a sentinel value like `const FOUND_IN_EXPORTS = {}` and use that as the symbol?

Comment (mischnic) [reply packages/core/core/src/BundleGraph.js:823]: You mean for `resolveSymbol(...).symbol`?
I would prefer an  ES6 `Symbol` or `false` in that case. Though `void | false | String` does play well with Flow's sketchy-null-string check.

**Pull request merged**